process/peer/peerTypeProvider.go

```diff
@@ -0,0 +1,198 @@
+	if check.IfNil(arg.NodesCoordinator) {
+		return nil, process.ErrNilNodesCoordinator
+	}
+	if check.IfNil(arg.EpochHandler) {
+		return nil, process.ErrNilEpochHandler
+	}
+	if check.IfNil(arg.ValidatorsProvider) {
+		return nil, process.ErrNilValidatorsProvider
+	}
+	if arg.EpochStartEventNotifier == nil {
+		return nil, process.ErrNilEpochStartNotifier
+	}
+	if arg.CacheRefreshIntervalDuration <= 0 {
+		return nil, process.ErrInvalidCacheRefreshIntervalDuration
+	}
+
+	ptp := &PeerTypeProvider{
+		nodesCoordinator:             arg.NodesCoordinator,
+		epochHandler:                 arg.EpochHandler,
+		validatorsProvider:           arg.ValidatorsProvider,
+		cache:                        make(map[string]*peerListAndShard),
+		mutCache:                     sync.RWMutex{},
+		cacheRefreshIntervalDuration: arg.CacheRefreshIntervalDuration,
+		refreshCache:                 make(chan bool),
+	}
+
+	ptp.startRefreshProcess()
+
+	arg.EpochStartEventNotifier.RegisterHandler(ptp.epochStartEventHandler())
+
+	return ptp, nil
+}
+
+// ComputeForPubKey returns the peer type for a given public key and shard id
+func (ptp *PeerTypeProvider) ComputeForPubKey(pubKey []byte) (core.PeerType, uint32, error) {
+	ptp.mutCache.RLock()
+	peerData, ok := ptp.cache[string(pubKey)]
+	ptp.mutCache.RUnlock()
+
+	// component is used so it could be refreshed on expiration
+	ptp.refreshCache <- false
+	if ok {
+		return peerData.pType, peerData.pShard, nil
+	}
+
+	return core.ObserverList, 0, nil
+}
+
+// GetAllPeerTypeInfos returns all known peer type infos
+func (ptp *PeerTypeProvider) GetAllPeerTypeInfos() []PeerTypeInfoHandler {
+	ptp.mutCache.RLock()
+	cache := ptp.cache
+	ptp.mutCache.RUnlock()
+
+	// component is used so it could be refreshed on expiration
+	ptp.refreshCache <- false
+
+	peerTypeInfos := make([]PeerTypeInfoHandler, 0, len(cache))
+	for pkString, peerListAndShard := range cache {
+		peerTypeInfos = append(peerTypeInfos, &PeerTypeInfo{
+			PublicKey: pkString,
+			PeerType:  string(peerListAndShard.pType),
+			ShardId:   peerListAndShard.pShard,
+		})
+	}
+
+	return peerTypeInfos
+}
+
+func (ptp *PeerTypeProvider) epochStartEventHandler() sharding.EpochStartActionHandler {
+	subscribeHandler := notifier.NewHandlerForEpochStart(func(hdr data.HeaderHandler) {
+		ptp.refreshCache <- true
+	}, func(_ data.HeaderHandler) {}, core.IndexerOrder)
+
+	return subscribeHandler
+}
+
+func (ptp *PeerTypeProvider) startRefreshProcess() {
+	go func() {
+		ptp.updateCache()
+		lastUpdate := time.Now()
+		for {
+			select {
+			case forceRefresh := <-ptp.refreshCache:
+				{
+					expired := time.Since(lastUpdate) > ptp.cacheRefreshIntervalDuration
+					if forceRefresh || expired {
+						lastUpdate = time.Now()
+						ptp.updateCache()
+					}
+				}
+			case <-time.After(ptp.cacheRefreshIntervalDuration):
+			}
+		}
+	}()
+}
+
+func (ptp *PeerTypeProvider) updateCache() {
+	allNodes, err := ptp.validatorsProvider.GetLatestValidatorInfos()
```

Comment:
How much does this function cost ?

Reply:
wouldn't it be better to refresh these caches let's say after X(100) commited blocks ?

Reply:
it depends how big the trie is and how much it is used. It could be updated every x blocks or x seconds without the optimization of usage.